packages/flutter/lib/src/widgets/basic.dart

```diff
@@ -2315,6 +2316,201 @@ class ConstrainedBox extends SingleChildRenderObjectWidget {
   }
 }
 
+/// A container widget that applies an arbitrary transform to its constraints,
+/// and sizes its child using the resulting [BoxConstraints], treating any
+/// overflow as error.
+///
+/// This container sizes its child using a [BoxConstraints] created by applying
+/// [constraintsTransform] to its own constraints. This container will then
+/// attempt to adopt the same size, within the limits of its own constraints. If
+/// it ends up with a different size, it will align the child based on
+/// [alignment]. If the container cannot expand enough to accommodate the entire
+/// child, the child will be clipped if [clipBehavior] is not [Clip.none].
+///
+/// In debug mode, if the child overflows the container, a warning will be
+/// printed on the console, and black and yellow striped areas will appear where
+/// the overflow occurs.
+///
+/// {@tool snippet}
+/// In the following snippet, the child [Card] is guaranteed to be at least as
+/// tall as its intrinsic height. Unlike [UnconstrainedBox], the child can be
+/// taller if the the parent's minHeight is higher than the child's intrinsic
+/// height. If parent container's maxHeight is less than [Card]'s intrinsic
+/// height, in debug mode a warning will be given.
+///
+/// ```dart
+/// ConstraintsTransformBox(
+///   constraintsTransform: (BoxConstraints constraints) => constraints.copyWith(maxHeight: double.infinity),
+///   child: const Card(child: Text('Hello World!')),
+/// )
+/// ```
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [ConstrainedBox], which renders a box which imposes constraints
+///    on its child.
+///  * [OverflowBox], a widget that imposes additional constraints on its child,
+///    and allows the child to overflow itself.
+///  * [UnconstrainedBox] which allows its children to render themselves
+///    unconstrained, expands to fit them, and considers overflow to be an error.
+class ConstraintsTransformBox extends SingleChildRenderObjectWidget {
+  /// Creates a widget that uses a function to transform the constraints it
+  /// passes to its child. If the child overflows the parent's constraints, a
+  /// warning will be given in debug mode.
+  ///
+  /// The [debugTransformType] argument adds a debug label to this widget.
+  ///
+  /// The `alignment`, `clipBehavior` and `constraintsTransform` arguments must
+  /// not be null.
+  const ConstraintsTransformBox({
+    Key? key,
+    Widget? child,
+    this.textDirection,
+    this.alignment = Alignment.center,
+    required this.constraintsTransform,
+    this.clipBehavior = Clip.none,
+    String debugTransformType = '',
+  }) : _debugTransformLabel = debugTransformType,
+       assert(alignment != null),
+       assert(clipBehavior != null),
+       assert(constraintsTransform != null),
+       super(key: key, child: child);
+
+  /// Creates a widget that imposes no constraints on its child, allowing it to
+  /// render at its "natural" size. If the child overflows the parent's
```

Comment:
What's the difference between "natural" as you use it here and "intrinsic" as in intrinsicHeight/Width?

Reply:
Also just out of curiosity, what if the child's intrinsic width or height is larger than the device's screen?  Will it actually be able to render that big or will it be constrained by the screen size?  (This would be larger than the parent in most normal cases, so it would also give an overflow warning, but I'm curious what would happen.)

Reply:
Ah I'm using them interchangeably. I'll replace intrinsic with natrual.

I think the child will be rendered as if there was infinite space (given it's not clipped) and will be aligned per its `alignment`. But yeah if it ends up being larger than the `RenderView` you'll get a warning (this is a debugging widget in some sense).

Reply:
I tried this out and for this app:

```dart
    MaterialApp(
      home: Scaffold(
        body: ConstraintsTransformBox(
          constraintsTransform: myConstraintsTransform,
          child: Container(
            color: Colors.red,
          ),
        ),
      ),
    );
```

When myConstraintsTransform is...

  * `(BoxConstraints constraints) => constraints,` the whole screen is red.
  * `(BoxConstraints constraints) => constraints.copyWith(maxHeight: 10000),` the whole screen is red, and there is an overflow warning.
  * `(BoxConstraints constraints) => constraints.copyWith(maxHeight: double.infinity),` the whole screen is empty, no warning.

Should there be an error for that last case with infinite height?  Or actually maybe that is the normal behavior when you make the constraints infinite in Flutter.

Reply:
`Container` has a [`LimitedBox`](https://master-api.flutter.dev/flutter/widgets/LimitedBox-class.html) which only kicks in when the size is unconstrained. I think what happened was when the maxHeight is unconstrained the `LimitedBox` will try to become as small as possible so the `ColoredBox` wrapped inside the `LimitedBox` is zero-sized (case 3).